artiq/compiler/transforms/inferencer.py

```diff
@@ -259,7 +259,31 @@ def visit_ExtSlice(self, node):
 
     def visit_SubscriptT(self, node):
         self.generic_visit(node)
-        if isinstance(node.slice, ast.Index):
+
+        if types.is_tuple(node.value.type):
+            if (not isinstance(node.slice, ast.Index) or
+                not isinstance(node.slice.value, ast.Num)):
+                diag = diagnostic.Diagnostic(
+                    "error", "tuples can only be indexed by a constant", {},
+                    node.slice.loc, []
+                )
+                self.engine.process(diag)
+                return
+
+            tuple_type = node.value.type.find()
+            index = node.slice.value.n
+            if index < 0 or index >= len(tuple_type.elts):
+                diag = diagnostic.Diagnostic(
+                    "error",
+                    "index {index} cannot be used to index tuple of size {size}",
```

Comment:
Nitpick, but I'd use "index … is out of range for tuple of size …", for consistent terminology with Python's default "IndexError: list index out of range".